File: Auto/car_app_library/showcase/src/main/java/com/google/android/libraries/car/app/samples/showcase/common/PlaceDetailsScreen.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.libraries.car.app.samples.showcase.common;

import static com.google.android.libraries.car.app.model.Action.BACK;

import android.content.Intent;
import android.net.Uri;
import androidx.annotation.NonNull;
import com.google.android.libraries.car.app.CarContext;
import com.google.android.libraries.car.app.Screen;
import com.google.android.libraries.car.app.model.Action;
import com.google.android.libraries.car.app.model.CarColor;
import com.google.android.libraries.car.app.model.Pane;
import com.google.android.libraries.car.app.model.PaneTemplate;
import com.google.android.libraries.car.app.model.Row;
import com.google.android.libraries.car.app.model.Template;
import java.util.ArrayList;
import java.util.List;

/** A screen that displays the details of a place. */
public class PlaceDetailsScreen extends Screen {
  private final PlaceInfo mPlace;

  public static PlaceDetailsScreen create(CarContext carContext, PlaceInfo place) {
    return new PlaceDetailsScreen(carContext, place);
  }

  @NonNull
  @Override
  public Template getTemplate() {
    List<Action> actions = new ArrayList<>();
    actions.add(
        Action.builder()
            .setTitle("Navigate")
            .setBackgroundColor(CarColor.BLUE)
            .setOnClickListener(this::onClickNavigate)
            .build());
    actions.add(
        Action.builder()
            .setTitle("Dial")
            .setOnClickListener(this::onClickDial)
            .build());

    Pane.Builder paneBuilder =
        Pane.builder()
            .setActions(actions)
            .addRow(Row.builder().setTitle("Address").addText(mPlace.address).build())
            .addRow(Row.builder().setTitle("Phone").addText(mPlace.phoneNumber).build());

    return PaneTemplate.builder(paneBuilder.build())
        .setTitle(mPlace.title)
        .setHeaderAction(BACK)
        .build();
  }

  private void onClickNavigate() {
    Uri uri = Uri.parse("geo:0,0?q=" + mPlace.address);
    Intent intent = new Intent(CarContext.ACTION_NAVIGATE, uri);
    getCarContext().startCarApp(intent);
  }

  private void onClickDial() {
    Uri uri = Uri.parse("tel:" + mPlace.phoneNumber);
    Intent intent = new Intent(Intent.ACTION_DIAL, uri);
    getCarContext().startCarApp(intent);
  }

  private PlaceDetailsScreen(CarContext carContext, PlaceInfo place) {
    super(carContext);
    mPlace = place;
  }
}

File: car_app_library/showcase/shared/src/main/java/androidx/car/app/sample/showcase/common/common/PlaceDetailsScreen.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.car.app.sample.showcase.common.common

import android.content.Intent
import android.net.Uri
import androidx.car.app.CarContext
import androidx.car.app.CarToast
import androidx.car.app.HostException
import androidx.car.app.Screen
import androidx.car.app.model.Action
import androidx.car.app.model.CarColor
import androidx.car.app.model.Pane
import androidx.car.app.model.PaneTemplate
import androidx.car.app.model.Row
import androidx.car.app.model.Template
import androidx.car.app.sample.showcase.common.R

/** A screen that displays the details of a place.  */
class PlaceDetailsScreen private constructor(
    carContext: CarContext,
    private val mPlace: PlaceInfo
) : Screen(carContext) {
    override fun onGetTemplate(): Template {
        val paneBuilder =
            Pane.Builder()
                .addAction(
                    Action.Builder()
                        .setTitle(carContext.getString(R.string.navigate))
                        .setBackgroundColor(CarColor.BLUE)
                        .setOnClickListener { this.onClickNavigate() }
                        .build())
                .addAction(
                    Action.Builder()
                        .setTitle(carContext.getString(R.string.dial))
                        .setOnClickListener { this.onClickDial() }
                        .build())
                .addRow(
                    Row.Builder()
                        .setTitle(carContext.getString(R.string.address))
                        .addText(mPlace.address)
                        .build()
                )
                .addRow(
                    Row.Builder()
                        .setTitle(carContext.getString(R.string.phone))
                        .addText(mPlace.phoneNumber)
                        .build()
                )

        return PaneTemplate.Builder(paneBuilder.build())
            .setTitle(mPlace.title)
            .setHeaderAction(Action.BACK)
            .build()
    }

    private fun onClickNavigate() {
        val uri = Uri.parse("geo:0,0?q=" + mPlace.address)
        val intent = Intent(CarContext.ACTION_NAVIGATE, uri)

        try {
            carContext.startCarApp(intent)
        } catch (e: HostException) {
            CarToast.makeText(
                carContext,
                carContext.getString(R.string.fail_start_nav),
                CarToast.LENGTH_LONG
            )
                .show()
        }
    }

    private fun onClickDial() {
        val uri = Uri.parse("tel:" + mPlace.phoneNumber)
        val intent = Intent(Intent.ACTION_DIAL, uri)

        try {
            carContext.startCarApp(intent)
        } catch (e: HostException) {
            CarToast.makeText(
                carContext,
                carContext.getString(R.string.fail_start_dialer),
                CarToast.LENGTH_LONG
            )
                .show()
        }
    }

    companion object {
        /** Creates an instance of [PlaceDetailsScreen].  */
        fun create(
            carContext: CarContext, place: PlaceInfo
        ): PlaceDetailsScreen {
            return PlaceDetailsScreen(carContext, place)
        }
    }
}
